renderlib/RotateTool.cpp

```diff
@@ -0,0 +1,367 @@
+        angle = sqrtf(yRadians * yRadians + xRadians * xRadians);
+        glm::vec3 objectUpDirection = camFrame.vy * yRadians;
+        glm::vec3 objectSidewaysDirection = camFrame.vx * xRadians;
+
+        glm::vec3 moveDirection = objectUpDirection + objectSidewaysDirection;
+        glm::vec3 eye = l - p;
+        glm::vec3 axis = glm::normalize(glm::cross(moveDirection, eye));
+
+        motion = glm::angleAxis(angle, axis);
+
+        break;
+    }
+
+    // The variable motion is a world space vector of how much we moved the
+    // manipulator. Here we execute the action of applying motion to
+    // whatever is that we are moving...
+    // [...]
+    // Note that the motion is the total motion from drag origin.
+    // but we will update this per-frame!
+    origins.rotate(scene, motion);
+    m_rotation = motion;
+  }
+  if (button.action == Gesture::Input::Action::kRelease) {
+    if (!origins.empty()) {
+      // Make the edit final, for example by creating an undo action...
+      // [...]
+    }
+
+    // Consume the gesture.
+    gesture.input.reset(Gesture::Input::kButtonLeft);
+    origins.clear();
+    m_rotation = glm::angleAxis(0.0f, glm::vec3(0, 0, 1));
+  }
+}
+
+void
+RotateTool::draw(SceneView& scene, Gesture& gesture)
+{
+  const glm::vec2 resolution = scene.viewport.region.size();
+  // Move tool is only active if something is selected
+  if (!scene.anythingActive()) {
+    return;
+  }
+
+  // Re-read targetPosition because object may have moved by the manipulator action,
+  // or animation. Target 3x3 linear space is a orthonormal frame. Target
+  // position is where the manipulator should be drawn in space
+  if (origins.empty()) {
+    origins.update(scene);
+  }
+  AffineSpace3f target = origins.currentReference(scene, Origins::kNormalize);
+
+  glm::vec3 viewDir = (scene.camera.m_From - target.p);
+  LinearSpace3f camFrame = scene.camera.getFrame();
+
+  // Draw the manipulator to be at some constant size on screen
+  float scale = length(viewDir) * scene.camera.getHalfHorizontalAperture() * (s_manipulatorSize / resolution.x);
+
+  AffineSpace3f axis;
+  axis.p = target.p;
+
+  gesture.graphics.addCommand(GL_TRIANGLES);
+  // draw a flat camera facing disk for freeform tumble rotation
+  float tumblescale = scale * 0.84f;
+  {
+    float opacity = 0.05f;
+    uint32_t code = manipulatorCode(RotateTool::kRotate, m_codesOffset);
+    glm::vec3 color = ManipColors::bright;
+    if (m_activeCode == RotateTool::kRotate) {
+      color = glm::vec3(1, 1, 0);
+      opacity = 0.2f;
+    }
+    gesture.drawCone(axis.p,
+                     camFrame.vy * tumblescale, // swapped x and y to invert the triangles to face user
+                     camFrame.vx * tumblescale,
+                     glm::vec3(0, 0, 0),
+                     48,
+                     color,
+                     opacity,
+                     code);
+  }
+  static constexpr uint32_t noCode = -1;
```

Comment:
Should this be moved to `Manipulator.h`?

Reply:
hm, yeah probably. There's actually a constant for it defined in the SelectionBuffer class. I should use that.